Render multiple plots

mistakevin opened the pull request:
Closes: #2348

mistakevin commented:
See comments and suggestion thread here https://github.com/bokeh/bokeh/pull/2394

bryevdv commented on bokeh/embed.py:
```diff
@@ -40,35 +40,78 @@ def components(plot_object, resources=None):
-    
+    all_models = []
+    plots = []
+    value_error = 'Input must be single or iterable collection of PlotObjects'
```
I think this would be more precise for users:
```
Input must be a PlotObject, a Sequence of PlotObjects, or a mapping of string to PlotObjects
```

bryevdv commented on sphinx/source/docs/user_guide/embed.rst:
```diff
@@ -119,6 +119,20 @@ For example, to use version ``0.8.2``:
+
+    components([foo, bar])               # (script, [foo_div, bar_div])
+
+    components((foo, bar))               # (script, (foo_div, bar_div))
```
Does it actually return a list if you pass in a list? I think any kind of sequence input, should always return a tuple output

mistakevin replied:
It returns the type of structure you put in, but I agree that it should return a tuple regardless.

mistakevin commented:
PR updated to include tests, minor improvements

bryevdv commented:
LGTM, @damianavila can you take a quick look as well and merge if you agree

mattpap commented:
What's going on with all those merged-in PRs?

bryevdv commented:
@mistakevin was working from a fork

damianavila commented:
Yes, already talk with him about that... I additionally add a tag branches to filter the noise...
BTW, I will merge this soon.

damianavila commented:
As a last requirement, I would like to see an example of multiple embedding. I think it would be useful for the user to see it in action.

mistakevin commented:
Updated Docs with a CodePen example

birdsarah commented:
The codepen doesn't run for me on Firefox - Blocked loading mixed active content "http://cdn.pydata.org/bokeh/release/bokeh-0.9.0.min.css"

Also - I'd love to help with / contribute an example that shows this off even more - I'm super excited about this feature. But I don't have time this week...maybe next week.

mistakevin commented:
Codepen loads for me on Firefox (Mac). Could it be maybe an add-on ad-blocker? I tested with Ad-Block Plus and Ghostery, CSS made it through both. Here is the [link](http://codepen.io/anon/pen/RPZYoq?editors=100) for anyone else who wants to test. Is Codepen a good option for this?

Also, it appears we need an account on Codepen to keep from randoms editing the code.

damianavila commented:
I can see the codepen here...

> Also, it appears we need an account on Codepen to keep from randoms editing the code.

We probably need to make an account... let's discuss this in the next meeting...

Others @bokeh/core can check if you can see the the codepen?

I want to merge this one today if possible...

bryevdv commented:
The pen works for me on Safari. There may already be a Bokeh Codeine account, check the password wallet.

mistakevin commented:
If we create an account after merging this, the link will probably change resulting in another PR to update the link in the docs. Should we just go ahead and make the account before the merge?

bryevdv commented:
So, for now, I think it would be best to add an actual `.py` example under `examples/plotting/file` that just saves out an HTML file by hand (instead of using `output_file` and `show`). I like the idea of putting some examples on Codepen but we can do that later after we get an account set up.

mistakevin commented:
@bryevdv
>that just saves out an HTML file by hand (instead of using `output_file` and `show`)

When the user runs `python embeddable_plots.py`:
Do we want it to output the built HTML file with the plots pre-inserted programatically? 
Or outputs the html boilerplate for the user to plug in the script and divs?

mistakevin commented:
@bryevdv @damianavila 
Docs and new example is now included. Removed codepen for now

fpliger commented on bokeh/embed.py:
```diff
@@ -40,35 +40,80 @@ def components(plot_object, resources=None):
+    from .plot_object import PlotObject
+    from .document import Document
+    from collections import Sequence
+    from six import string_types
```
are the local import an explicit need?

fpliger commented on bokeh/embed.py:
```diff
@@ -40,35 +40,80 @@ def components(plot_object, resources=None):
-            typically a Plot or PlotContext
+        plot_objects (PlotObject|list|dict|tuple) : Will return script
+        and div element to render if PlotObject, or correlating structures
+        of script/div combinations
```
not sure this description is straightforward to understand.. at least not for my not english-native brain. 

I'd also suggest to explain the difference (and benefits) of passing a list/tuple or a dict and add a small example.

fpliger commented:
Left a couple of comments. Thanks @mistakevin this is awesome in so many ways!

fpliger commented on embed.html:
```diff
@@ -0,0 +1,35 @@
+    <div class="plotdiv" id="8a571eab-100a-403d-b772-22bfc498d44f"></div><div class="plotdiv" id="6b798abf-c715-49d2-ae4d-bee9b2831e66"></div><div class="plotdiv" id="2ce8d176-a29d-41d1-b02c-fe34eb342ab5"></div>
+    </body>
+</html>
+    
```
Isn't this file the html generated by the new example? I don't think this belong to this PR..

bryevdv replied:
Weird I thought we had an Git ignore policy to ignore `.html` files under `examples`. Anyway, yah I agree we just want to check in the source `embed.py` and let users run it for themselves to generate the output file.

mistakevin commented:
Yep, that html file was an accidental check in, I've updated the comment with some output examples and a path to the embed.py example, and moved out the local imports.

damianavila commented:
@mistakevin, in examples we have a folder containing all the embed-related examples. I think you sould add this example there: https://github.com/bokeh/bokeh/tree/master/examples/embed

And you probably have to change the name from `embed` to somthing less generic... maybe, `multiple_embed.py`??

damianavila commented on examples/plotting/file/embed.py:
```diff
@@ -0,0 +1,73 @@
+
+script, div = components(plots)
+
+open('embed.html', 'w').close()
```
why `open` here?

mistakevin replied:
That's so if you tweak the code, it'll clear it before it writes to it again. Is there a better way to do this?

damianavila replied:
Just open it in write mode... and it will overwrite the previous file... https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files

something like this should work:

`with open("whatever.html", "w+") as f:`

bryevdv commented:
Last thing too: for the image tests to work the output filename has to match: e.g. `foo.py -> foo.html`

damianavila commented on examples/plotting/file/embed.py:
```diff
@@ -0,0 +1,73 @@
+
+import webbrowser
+import os
+try:
```
use `six`...`six.moves.urllib.request.pathname2url`

damianavila commented on examples/plotting/file/embed.py:
```diff
@@ -0,0 +1,73 @@
+script, div = components(plots)
+
+open('embed.html', 'w').close()
+with open("embed.html", "a") as textfile:
```
this would be nicer using jinja templates... somthing along this lines: http://nbviewer.ipython.org/github/bokeh/bokeh-notebooks/blob/master/tutorial/05%20-%20sharing.ipynb#Templating-in-HTML-Documents

mistakevin commented:
Moving `from .document import Document` to the top level causes an error in _document.py_ at `from .models import PlotContext`. Any thought on why that would happen?

damianavila commented:
Sorry, I have to be more clear... I just saying to move the file/embed.py example... not the embed.py module...

mistakevin commented:
Yeah, I understand that. But that error is still there, you can see where they failed the tests, and I can reproduce it locally.

mistakevin commented on bokeh/embed.py:
```diff
@@ -40,35 +43,91 @@ def components(plot_object, resources=None):
+        (script, div[s]): UTF-8 encoded
     '''
-    
+    from .document import Document
```
This is local because of a circular dependency problem I can't nail down. Any thoughts would be appreciated.

damianavila replied:
Well, this in interesting... pinging @bryevdv who probably can give us some hints on this...

mistakevin commented on bokeh/embed.py:
```diff
@@ -125,6 +184,10 @@ def file_html(plot_object, resources, title, template=FILE):
         html : standalone HTML document with embedded plot
 
     '''
+    from .document import Document
```
Ditto here

damianavila commented:
OK, I think this is ready, I will merge it later today...
